### PR TITLE
Remove map file if it was created with the mapping

### PR DESCRIPTION
**Pull request by tobim**
The behavior mmapbuf should be symmetric, i.e. if the underlying file did not exist before creating the memory map, the file should be deleted when the mapping is removed.

### CODE REVIEW

File: libvast/src/detail/mmapbuf.cpp [hunk 1]
```diff
@@ -83,6 +85,8 @@ mmapbuf::~mmapbuf() {
     munmap(map_, size_);
   if (fd_ != -1)
     close(fd_);
+  if (is_file_owner_)
+    std::remove(filename_.str().c_str());
```
Comment (Neverlord) [inline libvast/src/detail/mmapbuf.cpp]: Why do we need `c_str` here?`std::path` has a constructor that takes a string: https://en.cppreference.com/w/cpp/filesystem/path/path, or do I miss something?

Comment (mavam) [reply libvast/src/detail/mmapbuf.cpp]: @Neverlord We haven't converted `vast::path` to `std::path` yet, perhaps that's the reason.

My comment here would be: we haven't used the new filesystem API so far in VAST, this is the first use. I'd rather use the old stuff from `vast/filesystem.hpp` (here: `vast::rm`) and then migrate all filesystem operations at once. Mixing and matching results only in confusion.

File: libvast/src/detail/mmapbuf.cpp [hunk 2]
```diff
@@ -57,6 +57,8 @@ mmapbuf::mmapbuf(const path& filename, size_t size, size_t offset)
     file_size = st.st_size;
   else if (result < 0 && errno != ENOENT)
     return;
+  else
+    is_file_owner_ = true;
```
Comment (mavam) [inline libvast/src/detail/mmapbuf.cpp]: This check auto-detects the file size. It's not intuitive why an `else` branch implies that the user owns the file. I suggest clarifying this with a comment or spliting the logic.

Comment (tobim) [reply libvast/src/detail/mmapbuf.cpp]: > I'm not sure if I understand the implications of this PR. When and why do we want to automatically delete a file that we created? Shouldn't deletion of a file be something the user decides after having closed the streambuffer?

In general terms: Any object that allocates a resource should release that resource at the end of its lifetime. Why should that only apply to memory?
But your question prompted me to think more about the interface to `mmapbuf` and I think a better approach might be to have two sparate constructors, one for owning an the other for non-owning `mmapbuf`s. I'll append another commit to this PR for the change.

> This check auto-detects the file size. It's not intuitive why an else branch implies that the user owns the file. I suggest clarifying this with a comment or spliting the logic.

You are quite correct.

File: libvast/src/detail/mmapbuf.cpp [hunk 3]
```diff
@@ -45,29 +45,42 @@ mmapbuf::mmapbuf(size_t size)
   setp(map_, map_ + size_);
 }
 
+mmapbuf::mmapbuf(const int fd, const size_t size, const size_t offset)
+  : is_file_owner_{true},
```
Comment (mavam) [inline libvast/src/detail/mmapbuf.cpp]: Why does the overload that passes a file descriptor imply ownership? Shouldn't it be the opposite? In this version, you're calling `std::remove` in the destructor, which doesn't make much sense to me.

File: libvast/test/mmapbuf.cpp [hunk 4]
```diff
@@ -34,31 +39,36 @@ TEST(memory-mapped streambuffer) {
...
+    n = sb.sputn(data.data(), data.size());
+    CHECK_EQUAL(n, static_cast<std::streamsize>(data.size()));
+    CHECK_EQUAL(std::string(sb.data(), sb.size()), "corge bazqux");
+    CHECK(sb.resize(data.size() - 1));
+    // Figure out current position.
+    size_t cur = sb.pubseekoff(0, std::ios::cur, std::ios::out);
+    CHECK_EQUAL(cur, sb.size()); // we're at the end!
+  }
+  close(fd);
+  std::remove(filename.str().c_str());
```
Comment (mavam) [inline libvast/test/mmapbuf.cpp]: Per my previous review, please use `vast::rm` in favor of `std::remove` until we have switched to the C++17 filesystem API.

File: libvast/src/detail/mmapbuf.cpp (hunk 3 above)
Comment (tobim) [reply libvast/src/detail/mmapbuf.cpp]: This is a mistake, thanks for pointing it out.

**Comment by tobim** [comment]
This PR was a bit overzealous I'm afraid. The idea here was that an object should not leave things (files) around when it created them (RAII-style). What I didn't consider is that RAII only applies to resources that are definitely lost if one drops all pointers to it, but files can still be reached through the filesystem, so its kinda pointless (pun intended).

The original motivation was because of the unit tests that left files lying around, cleaning that up should probably be done by the fixture instead.

I will withdraw this PR then.

**Comment by mavam** [comment]
Okay, I'm glad we ended up on the same page.

**Pull request closed without merging**